### DISCUSSION TITLE
Troubleshooting body part symmetry

### DISCUSSION THREAD

**Discussion by amblypatty** [talmolab/sleap#1573] [Help!]
Hello SLEAP,

I am having trouble getting a UNET model to train on a skeleton that has symmetrical body parts, from a bird's eye view, very similarly to the fly32 skeleton. I am using a top-down model for a single-instance (arachnid) with thin legs in an arena (fig. 1). 
![ws_in_arena](https://github.com/talmolab/sleap/assets/72351141/7db5f394-68c6-4e38-ba9e-aba8ad02acfe)
(Fig 1. Whip Spider in arena from video frame) 

![ws_full_skeleton](https://github.com/talmolab/sleap/assets/72351141/e384a647-a0bf-4510-a8b2-aba72ef7f34d)
(Fig 2. Full 35 keypoint, 26 edge, skeleton in an annotated frame)

The centroid model trains just fine, but the centered_instance model is what I cannot get to learn to avoid mis-crossing of symmetrical keypoints. The skeleton identifies symmetry (fig. 2), so I don't understand why this is still a problem for the model. The nodes that get mis-crossed NEVER cross over each other in the real video. I have tried increasing the number of filters (from 48-64-96), increasing the node sigma (from 2.0-2.5-5.0), doubling the number of annotated training frames (from 250-500), adjusting the online hard keypoint mining min and max (from min=2 to min=8, max=8 to max=28), increasing the batch size (from 4-8), resuming training for more time (another 200 epochs with early stopping enabled at plateau of 1e-8), lowering the initial learning rate (from 1e-4 to 2.5e-5).  The main problem of symmetry on some body parts still appears to not be correct (fig. 3), yet the model says it trained well (fig. 4). 
![ws_prediction_symm_error](https://github.com/talmolab/sleap/assets/72351141/efd7a1b6-d044-4858-a02d-2d01f4e0b449)
(Fig 3. The red arrow shows an erroneous crossed body part on the instance. Note the mean score of 0.83 in red.)

![training_good_OKSmAP](https://github.com/talmolab/sleap/assets/72351141/64abdeb4-9731-499e-8e5b-f0bcf170edcd)
(Fig 4. The model states that both the training and validation sets were more than 95% accurate, but I would expect the instances that predict crossed body parts should generate enormous distance errors for the model.)

How many training annotations are needed, or which training parameters should I adjust to get rid of this issue? I find it boggling because the mean score for some of the erroneous instances are _higher_ than the mean score for some instances that have no errors (figs. 3 & 5)! 
![ws_prediction_no_symm_error](https://github.com/talmolab/sleap/assets/72351141/7b45162e-c001-4c1a-a1e2-15f4f82da1cf)
(Fig 5. A predicted instance from the same training set that predicted all body parts correctly. Note the score of 0.77 in red is lower than the mean score from fig. 3.)

I think this has to do with the model's confidence about the predictions, but this metric is useless if I want to isolate erroneous instances to correct them.

Alternatively, is there a better way to check if there are proportionally large jumps from frame to frame for each node based on some standard deviations from the mean of the previous 10-150 frames? Perhaps if I were able to isolate all of the erroneous nodes, I would be able to correct them via interpolation before exporting the pose data for analysis.

Any help on this front would be appreciated. Please let me know if any more information could help solve my problem. I have been working on this for months and am desperate to get this project to the analysis stage. 

Best,
Patrick

**Comment by roomrys** [comment]
Hi @amblypatty,

### Body bart symmetry
The body part symmetry is used to augment the data for training. This would basically flip the labels on your skeleton (based on the symmetrical node) with the assumption that those nodes have similar image statistics to create more labels from less. Body part symmetry does nothing to ensure left/right nodes do not swap.

### Param to adjust
As you stated, the scored indeed are a measure of confidence of how well SLEAP thinks it did. If I had to guess why SLEAP thinks it did so well, it would be because the labeling along the legs all look pretty similar and perhaps the receptive field (the amount of the image that the network sees at a single time) was not big enough to know that the antlegR5's have 3 "edges" which are background and one which is foreground whereas all other antlegs always only have two fore/background "edges". Although centered-instance does not support adjusting the input scaling (implemented s.t. cropped image always retains full resolution), to adjust the RF, you can play with the Max Stride.

### Other examples? (to brainstorm)
If you have more examples of the body part swaps, I could help brainstorm ideas for why the network might think that is the best prediction of body parts. In you provided example, it seems antleg3 was missed and perhaps that is what led to antleg4 being placed in the incorrect spot... although I think we should try RF first as you already stated that you've increased labeling.

### Idea for determining swaps
Your idea of having some way of finding the swapped nodes (not relying on the score which is proving to have a serious case of false confidence), made me think of an idea I had a bit back (probably a too complicated idea and there is definitely an easier solution)... Is there a way we could draw a line(s) to represent symmetry and know that a node is switched using this line of symmetry. We only know symmetry between nodes and we aren't explicitly sure which direction that symmetry is expected in relation to other body parts, but this is something that we can gather statistics of and use those statistics to determine outliers. Maybeee using a voronoi diagram to draw these regions/lines-of-symmetry and assigning each region a probability of being a different symmetry (if that node is symmetrical). If you already have predictions on all frames, then I think your idea will be much easier to implement (and debug).
- #1582

Thanks,
Liezl

**Reply by amblypatty** [reply]
Hi @roomrys,

Thanks for the reply and adding the idea for finding swapped nodes. Indeed, I did not know that body part symmetry was only used for augmentation rather than keeping control during training of the swapping issue. In my case with my skeleton, before I input the symmetry for the nodes, I wasn't able to train anything except for the central three nodes. I even deleted all symmetrical nodes except for two that I expected would be easy for the model to learn, but it still would not learn them well. I would explain in the documentation/tutorials how critical symmetry is to identify in the skeleton for skeletons with symmetrical nodes. 

I will try increasing the RF by increasing the max stride and see if that helps. I will post updates here.

For other examples of swaps for brainstorming, I will share a recent training package and the predictions made for it after training. Any other ideas for why the model is swapping them and how to get it to learn to avoid swapping would be of great help.

Best,
Patrick

**Reply by amblypatty** [reply]
Hi @roomrys,

I finally got the training to complete with doubling the max stride (from 16 to 32) on Google Colab. *Side rant, Google Colab and Google Drive have been more difficult to work with than they should be as GCP times out after 2 hrs and now GD has been unmounting randomly in the middle of training, causing the cell to keep running and wasting compute units. So, I've had to resume training three times just to finish a single training session which costs almost $15 for the A100 GPU (necessary for the large model).* 

Below are the model hyperparameters and the metrics for the 16 max stride and the 32 max stride models. With the first model, for the 116 suggested frames I predicted on after training, there were 20 erroneous frames that had at least one visible node missing, swapped, or significantly misplaced (>20px). In the second model, only two of the erroneous instances were corrected. For some background, the camera took 1080p video at 60 Hz from only 0.3 m above the subjects. Videos were reencoded with ffmpeg with a crf of 15 on slow for nearly lossless encoded videos. Do the images really look so grainy that some nodes can't be uniquely learned? Or could the "hard-to-learn" nodes (like the antleg4s) just have too much variability for the model to learn them? I might have to delete the nodes but am afraid that the model will just find two other visible nodes that it thinks are difficult to find. 

16 Max Stride
[centered_instance.json](https://github.com/talmolab/sleap/files/13255965/centered_instance.json)
![231031_100518 centered_instance_metrics](https://github.com/talmolab/sleap/assets/72351141/a294558d-dc4f-468a-b3cf-c695d86e437c)


32 max stride
[centered_instance.json](https://github.com/talmolab/sleap/files/13255971/centered_instance.json)
![231103_021343 centered_instance_metrics](https://github.com/talmolab/sleap/assets/72351141/f3cc04cd-a4e3-4e2e-953a-6401d1bb43a4)

Some examples of erroneous instances:
![231103_021343 centered_instance_error_frame523-1-27616](https://github.com/talmolab/sleap/assets/72351141/b91771e5-178e-4b43-b8c3-eb6a720c3d95)
![231103_021343 centered_instance_error_frame551-15-22052](https://github.com/talmolab/sleap/assets/72351141/2986f721-34d0-446f-8338-1de2ba176e47)
![231103_021343 centered_instance_error_frame557-18-12224](https://github.com/talmolab/sleap/assets/72351141/74fcac96-76b5-4ebe-bec5-3bf4161630db)
![231103_021343 centered_instance_error_frame566-22-13649](https://github.com/talmolab/sleap/assets/72351141/a1a4de20-1882-4611-9ba5-dd1f0c6de578)

Any insight would be appreciated.

Best,
Patrick

**Reply by amblypatty** [reply]
Hi @roomrys ,

Another update, as I was looking at visualizations during the last training session, I saw that one of the antleg5 points appeared to be cut off by the cropping function of the centered_instance parameter. So, I increased the crop size from 640 to 704 (the red box) to include all the points in all the instances (and divisible by the max stride) and I saw another slight increase in the training accuracy (three more predictions were corrected). 

![Largest_instance_from_pedicel_crop](https://github.com/talmolab/sleap/assets/72351141/ecb473b4-f5f4-4041-a122-66a74a049f8d)

As I was looking at SLEAP's model evaluation notebook, the example has an n=1800 frames. While my pixel accuracy and OKS appears to be very good compared to the example data, my guess is that I am just going to have to increase the number of annotated training frames to somewhere near example before I can get rid of the node swapping issue.

![231101_174755 centered_instance_localization_error_graph](https://github.com/talmolab/sleap/assets/72351141/1a65d53f-eb91-4187-bfc6-f7a2327a35af)
![231101_174755 centered_instance_OKS_graph](https://github.com/talmolab/sleap/assets/72351141/c9eef34e-d585-44a1-90c0-b77ad78185d2)

Again, if you have any other insight from the files I sent you, please let me know, and also please let me know if you think I am on the right track.

Best,
Patrick
